azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -27,12 +55,44 @@ class PolicyError(AgentError):
     """
 
 
-class PolicyEngine(object):
+class InvalidPolicyError(AgentError):
+    """
+    Error raised if user-provided policy is invalid.
+    """
+    def __init__(self, msg, inner=None):
+        msg = "Customer-provided policy file ('{0}') is invalid, please correct the following error: {1}".format(conf.get_policy_file_path(), msg)
+        super(InvalidPolicyError, self).__init__(msg, inner)
+
+
+class _PolicyEngine(object):
     """
     Implements base policy engine API.
     """
-    @classmethod
-    def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_event=True):
+    def __init__(self):
+        """
+        _PolicyEngine should be initialized on a per-goal state basis. Policy enablement is checked and policy file is
+        read only during initialization. The same policy is enforced even if the policy file is deleted or changed
+        during a single goal state processing.
+        """
+        self._policy_enforcement_enabled = self.__get_policy_enforcement_enabled()
+        if not self.policy_enforcement_enabled:
+            return
+
+        # Set defaults for policy and update with the customer-provided policy file.
+        self._policy = \
+        {
+            "policyVersion": _MAX_SUPPORTED_POLICY_VERSION,
+            "extensionPolicies": {
+                "allowListedExtensionsOnly": _DEFAULT_ALLOW_LISTED_EXTENSIONS_ONLY,
+                "signatureRequired": _DEFAULT_SIGNATURE_REQUIRED,
+                "extensions": {}
+            }
+        }
+        custom_policy = self.__get_policy()
+        self.__parse_policy(custom_policy)
```

Comment:
__parse_policy may update self._policy with a partially parsed policy in the case that an exception is raised during parsing. 

I don't think self._policy should be in a 'partially parsed' state. For now, I understand why this may be ok, but it could be risky in future iterations of the code.

I think self._policy should either be -> default policy, or fully parsed custom policy

Reply:
Based on Maddie's comments, I changed the implementation to update a template variable, and then set self._policy = template at the end of init. Let me know if this implementation makes sense.